File: sgfpy/goboard.py
### goboard.py provides the board and move models.  The main classes are
### GoBoard, Move, and Adornments.

import clr

## Debugging only ...
## Needed for System.Windows
#clr.AddReference('PresentationFramework')
#from System.Windows import MessageBox

__all__ = ["GoBoard", "Move", "Adornments", "parsed_move_model_coordinates", 
           "parsed_label_model_coordinates", "parsed_to_model_coordinates"]


### GoBoard models the board in terms of what stones are where on the
### board and has a list/tree of Moves.  The board model is that rows
### increase going down from the top, and columns increase to the right.
###
class GoBoard (object):
    
    supported_sizes = [9, 13, 19]
    
    def __init__ (self, size):
        ## size holds the number of lines on the board.
        if size not in GoBoard.supported_sizes:
            raise Exception("Board size must be 9, 13, or 19 -- got " + str(size))
        self.size = size
        ## moves holds Move objects or None if there's no stone at the location
        self.moves = [[None for col in xrange(size)] for row in xrange(size)]

    ### add_stone adds move to the model, assuming it has valid indexes.
    ### row, col are one-based, as we talk about go boards.
    ###
    def add_stone (self, move):
        if self.moves[move.row - 1][move.column - 1] is not None:
            raise Exception("Ensure board has no stone at location.")
        self.moves[move.row - 1][move.column - 1] = move
        return move
    
    ### remove_stone removes the move indicated by move's row and column.
    ### There is no identity check that the argument is in the model.  Row, col
    ### are one-based, and this assumes they are valid indexes.
    ###
    def remove_stone (self, move):
        self.moves[move.row - 1][move.column - 1] = None
    
    def remove_stone_at (self, row, col):
        if self.move_at(row, col):
            self.remove_stone(self.move_at(row, col))
    
    ### goto_start removes all stones from the model.
    ###
    def goto_start (self):
        for row in xrange(self.size):
            for col in xrange(self.size):
                self.moves[row][col] = None
    
    ### move_at returns the move at row, col (one-based indexes), or None if
    ### there is no move here.
    ###
    def move_at (self, row, col):
        return self.moves[row - 1][col - 1]
    
    ### color_at returns the color of the stone at row, col (one-based), or None if there is not stone there.
    ### This assumes row, col are valid indexes.
    ###
    def color_at (self, row, col):
        m = self.moves[row - 1][col - 1]
        if m:
            return m.color
        else:
            return None


    ### has_stone returns true if the go board location row,col (one-based) has
    ### a stone.  This function assumes row and col are valid locations.
    ###
    def has_stone (self, row, col):
        return self.move_at(row, col) is not None


    ### The following functions return whether the specified location has a
    ### stone or a stone of a particular color at the specified location.  The
    ### row, col are one-based, and if an index is invalid, the functions
    ### return false so that the edges of the board never have a stone
    ### essentially.

    def has_stone_left (self, row, col):
        return ((col - 1) >= 1) and (self.move_at(row, col - 1) is not None)
    
    def has_stone_color_left (self, row, col, color):
        return self.has_stone_left(row, col) and (self.move_at(row, col - 1).color == color)
    
    
    def has_stone_right (self, row, col):
        return ((col + 1) <= self.size) and (self.move_at(row, col + 1) is not None)
    
    def has_stone_color_right (self, row, col, color):
        return self.has_stone_right(row, col) and (self.move_at(row, col + 1).color == color)
    
    
    def has_stone_up (self, row, col):
        return ((row - 1) >= 1) and (self.move_at(row - 1, col) is not None)
    
    def has_stone_color_up (self, row, col, color):
        return self.has_stone_up(row, col) and (self.move_at(row - 1, col).color == color)
    
    
    def has_stone_down (self, row, col):
        return ((row + 1) <= self.size) and (self.move_at(row + 1, col) is not None)
    
    def has_stone_color_down (self, row, col, color):
        return self.has_stone_down(row, col) and (self.move_at(row + 1, col).color == color)
    


### Move models a move or stone on the board and links to the previous
### and next moves.
###
class Move (object):
    
    def __init__ (self, row, column, color):
        ## Row and column are None when this is a pass move.
        self.row = row
        self.column = column
        self.is_pass = row is None and column is None
        self.color = color
        ## The following need to be set if placing move into a game
        self.next = None
        self.previous = None
        self.number = 0
        self.dead_stones = []
        ## Adornments is for editing markup.
        self.adornments = []
        ## Comments holds text describing the move or current game state.
        self.comments = ""
        ## Branches holds all the next moves, while next points to one of these.
        ## This is None until there's more than one next move.
        self.branches = None
        ## parsed_node is a node from a .sgf file.
        self.parsed_node = None
        ## rendered by default is true assuming moves are made and immediately
        ## displayed, but parsed nodes can have unprocessed branches or
        ## annotations.
        self.rendered = True
    
    
    ### The adornment functions do no checking on the objects added, whether
    ### objects are really in the collection for removing, etc.  It's up to
    ### users to use them correctly.
    
    def add_adornment (self, a):
        self.adornments.append(a)
        return a
    
    def remove_adornment (self, a):
        self.adornments.Remove(a)

###
### Coordinates conversions
###

### letters used for translating parsed coordinates to get model coordinates.
###
_letters = ["skip_0_goboard_one_based", 
            "a", "b", "c", "d", "e", "f", "g", "h", "i", "j",
            "k", "l", "m", "n", "o", "p", "q", "r", "s"]

### get_parsed_coordinates returns letter-based coordinates from _letters for
### writing .sgf files.  If flipped, then return the diagonal mirror image
### coordinates for writing files in the opponent's view of the board.
###
def get_parsed_coordinates (move_or_adornment, flipped):
    if type(move_or_adornment) is Move and move_or_adornment.is_pass:
        return ""
    if flipped:
        return (_letters[20 - move_or_adornment.column] +
                _letters[20 - move_or_adornment.row])
    else:
        return (_letters[move_or_adornment.column] +
                _letters[move_or_adornment.row])

def flip_parsed_coordinates (coords):
    coords = parsed_to_model_coordinates(coords)
    if None in coords:
        return ""
    else:
        return _letters[20 - coords[1]] + _letters[20 - coords[0]]
    


### parsed_label_model_coordinates takes a parsed properties and returns as
### multiple values the row, col (in terms of the model used by goboard.py),
### and the label letter.  Data should be "<letter><letter>:<letter>".
###
def parsed_label_model_coordinates (data):
    return parsed_to_model_coordinates(data) + (data[3],)
    
### parsed_to_model_coordinates takes a parsed coordinates string and returns
### as multiple values the row, col in terms of the model used by goboard.py.
### This assumes coords is "<letter><letter>" and valid indexes.
###
def parsed_to_model_coordinates (coords):
    if coords == "":
        ## Pass move
        return None, None
    else:
        return (_letters.index(coords[1]), _letters.index(coords[0]))

###
### Adornments
###

### Adornments is probably misnamed, but instances represent adornments to
### locations on the go board, such as triangles, squares, letters, etc., that
### let comments identify moves by moves and locations by more than
### coordinates.  The class object has "static" members to represent the
### constant kinds of adornments.  There are some helpes to manage the unique
### current move adornment.
###
class Adornments (object):
    
    ### __init__ takes a kind, which is a unique instance of Adornments set
    ### below after the class.  It also takes the go board location
    ### (one-based).  If move is supplied, then the new adornment is on a move
    ### that is on the board at row, col.
    ###
    def __init__ (self, kind, row, col, move = None, letter = None):
        ## Kind is static member of class set below.
        self.kind = kind
        ## Letter is string when kind is letter.
        self.letter = letter
        self.row = row
        self.column = col
        self.move = move
        ## Cookie is for the UI layer to hold onto objects per adornment.
        self.cookie = None

    ## There's only one current move adornment.
    current_move_adornment = None

    @staticmethod
    def get_current_move (move, cookie):
        if Adornments.current_move_adornment.move is not None:
            raise Exception("Already have current move adornment at row, col: " +
                            str(Adornments.current_move_adornment.move.row) + ", " +
                            str(Adornments.current_move_adornment.move.column) + ".")
        Adornments.current_move_adornment.move = move
        Adornments.current_move_adornment.cookie = cookie
        move.add_adornment(Adornments.current_move_adornment)
    
    @staticmethod
    def release_current_move ():
        move = Adornments.current_move_adornment.move
        if move is None:
            raise Exception("Do not have current move adornment.")
        move.adornments.remove(Adornments.current_move_adornment)        
        Adornments.current_move_adornment.move = None
        Adornments.current_move_adornment.cookie = None

### Kinds of adornments ...
Adornments.triangle = object()
Adornments.square = object()
Adornments.letter = object()
Adornments.current_move = object()
### The unique current move adornment
Adornments.current_move_adornment = Adornments(Adornments.current_move, 1, 1)

File: sgfedwin10/GoBoard.cs
//// goboard.cs provides the board and move models.  The main classes are
//// GoBoard, Move, and Adornments.

using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml; // UIElement
// System.Windows: UIElement, MessageBox
using System.Windows;
//using System.Linq;
//using System.Text;
using Microsoft.UI; // Color, Colors
using Color = Windows.UI.Color;
//using System.Windows.Media; // Color, Colors
using System.Diagnostics; // Debug.Assert


namespace SgfEdwin10 {

    public class GoBoard {
        public int Size { get; set; }
        private Move[,] moves = null;

        public GoBoard (int size) {
            this.Size = size;
            this.moves = new Move[size, size];
        }


        //// add_stone adds move to the model, assuming it has valid indexes.
        //// row, col are one-based, as we talk about go boards.
        ////
        public Move AddStone (Move move) {
            // This debug.assert could arguably be a throw if we think of this function
            // as platform/library.
            MyDbg.Assert(this.moves[move.Row - 1, move.Column - 1] == null,
                         "Ensure board has no stone at location.");
            this.moves[move.Row - 1, move.Column - 1] = move;
            return move;
        }

        //// remove_stone removes the move indicated by move's row and column.
        //// There is no identity check that the argument is in the model.  Row, col
        //// are one-based, and this assumes they are valid indexes.
        ////
        public void RemoveStone (Move move) {
            this.moves[move.Row - 1, move.Column - 1] = null;
        }

        public void RemoveStoneAt (int row, int col) {
            if (this.MoveAt(row, col) != null)
                this.RemoveStone(this.MoveAt(row, col));
        }

        //// goto_start removes all stones from the model.
        ////
        public void GotoStart () {
            for (var row = 0; row < this.Size; row++)
                for (var col = 0; col < this.Size; col++)
                    this.moves[row, col] = null;
        }

        //// move_at returns the move at row, col (one-based indexes), or None if
        //// there is no move here.
        ////
        public Move MoveAt (int row, int col) {
            return this.moves[row - 1, col - 1];
        }

        //// BoardModelAsSTring returns a string to lightly display board state for debugging purposes.
        ////
        public string BoardModelAsString () {
            var res = "";
            for (var row = 0; row < this.Size; row++) {
                for (var col = 0; col < this.Size; col++)
                    if (this.moves[row, col] != null) {
                        if (this.moves[row, col].Color == Colors.White) {
                            res = res + "O";
                        }
                        else if (this.moves[row, col].Color == Colors.Black) {
                            res = res + "X";
                        }
                        else
                            res = res + "?";
                    }
                    else
                        res = res + "+";
                res = res + "\n";
            }
            return res;
        }

        //// color_at returns the color of the stone at row, col (one-based), or None if there is not stone there.
        //// This assumes row, col are valid indexes.
        ////
        public Color ColorAt (int row, int col) {
            var m = this.moves[row - 1, col - 1];
            if (m != null)
                return m.Color;
            else
                return GoBoardAux.NoColor;
        }


        //// has_stone returns true if the go board location row,col (one-based) has
        //// a stone.  This function assumes row and col are valid locations.
        ////
        public bool HasStone (int row, int col) {
            return this.MoveAt(row, col) != null;
        }

        //// The following functions return whether the specified location has a
        //// stone or a stone of a particular color at the specified location.  The
        //// row, col are one-based, and if an index is invalid, the functions
        //// return false so that the edges of the board never have a stone
        //// essentially.

        public bool HasStoneLeft (int row, int col) {
            return ((col - 1) >= 1) && (this.MoveAt(row, col - 1) != null);
        }

        public bool HasStoneColorLeft (int row, int col, Color color) {
            return this.HasStoneLeft(row, col) && (this.MoveAt(row, col - 1).Color == color);
        }


        public bool HasStoneRight (int row, int col) {
            return ((col + 1) <= this.Size) && (this.MoveAt(row, col + 1) != null);
        }

        public bool HasStoneColorRight (int row, int col, Color color) {
            return this.HasStoneRight(row, col) && (this.MoveAt(row, col + 1).Color == color);
        }


        public bool HasStoneUp (int row, int col) {
            return ((row - 1) >= 1) && (this.MoveAt(row - 1, col) != null);
        }

        public bool HasStoneColorUp (int row, int col, Color color) {
            return this.HasStoneUp(row, col) && (this.MoveAt(row - 1, col).Color == color);
        }


        public bool HasStoneDown (int row, int col) {
            return ((row + 1) <= this.Size) && (this.MoveAt(row + 1, col) != null);
        }

        public bool HasStoneColorDown (int row, int col, Color color) {
            return this.HasStoneDown(row, col) && (this.MoveAt(row + 1, col).Color == color);
        }

    } // GoBoard class



    //// GoBoardAux provides stateless helper functions for GoBoard.  Public members are used
    //// outside GoBoard, while internal members are only used within GoBoard.
    ////
    public static class GoBoardAux {

        //// NoColor is necessary since Color vars cannot be null;
        ////
        internal static Color NoColor = default(Color);

        //// NoIndex represents coordinates for pass moves.
        ////
        internal static int NoIndex = -100;

        ////
        //// Coordinates Conversions
        ////

        //// Letters used for translating parsed coordinates to model coordinates.
        //// The first element is bogus because the model is 1 based to match user model.
        ////
        private static char[] letters = {((char)0), 
                                         'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j',
                                         'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's'};

        //// get_parsed_coordinates returns letter-based coordinates from _letters for
        //// writing .sgf files.  If flipped, then return the diagonal mirror image
        //// coordinates for writing files in the opponent's view of the board.
        ////
        //// SGF format is col,row (count from left edge, count from top).
        ////
        //// UWP runtime completely screws over dynamic, and while this function didn't take 4-6s
        //// like some, it just seems prudent to complicate the code with two definitions to guard against
        //// more lossage later (like they make dynamic actually not work).  And sure enough, by 2018,
        //// they screwed the pooch so bad on dynamic by throwing away the inline caching and interpreting
        //// all dynamic code such that function calls took several seconds instead of micro-seconds.
        ////
        public static string GetParsedCoordinates (Move move_or_adornment, bool flipped, int size) {
            int row = move_or_adornment.Row;
            int col = move_or_adornment.Column;
            if (move_or_adornment.IsPass)
                return "";
            if (flipped)
                // C# fails mutually distinct types on char and int, and string constructor lame.
                // Must cons array so that string constructor doesn't interpret second char as count.
                return new string(new char[] {letters[size + 1 - col], letters[size + 1 - row]});
            else
                return new string(new char[] {letters[col], letters[row]});
        }
        public static string GetParsedCoordinatesA (Adornments move_or_adornment, bool flipped, int size) {
            int row = move_or_adornment.Row;
            int col = move_or_adornment.Column;
            if (flipped)
                // C# fails mutually distinct types on char and int, and string constructor lame.
                // Must cons array so that string constructor doesn't interpret second char as count.
                return new string(new char[] { letters[size + 1 - col], letters[size + 1 - row] });
            else
                return new string(new char[] { letters[col], letters[row] });
        }

        //// FlipParsedCoordinates is only called from GameAux.FlipCoordinates, which is only used for
        //// writing files.  Hence, we represent pass coordinates as the empty string.  Apparently, some
        //// programs use "tt" due to older format conventions.
        ////
        public static string FlipParsedCoordinates(string coords, int size) {
            var mcoords = GoBoardAux.ParsedToModelCoordinates(coords);
            if (mcoords.Item1 == GoBoardAux.NoIndex)
                return "";
            else
                // C# fails mutually distinct types on char and int, and string constructor lame.
                // Must cons array so that string constructor doesn't interpret second char as count.
                return new string(new char[] { letters[size + 1 - mcoords.Item2], letters[size + 1 - mcoords.Item1] });
        }
    

        //// parsed_label_model_coordinates takes a parsed properties and returns as
        //// multiple values the row, col (in terms of the model used by goboard.py),
        //// and the label letter.  Data should be "<letter><letter>:<letter>".
        ////
        public static Tuple<int, int, char> ParsedLabelModelCoordinates (string data) {
            var tmp = GoBoardAux.ParsedToModelCoordinates(data);
            return new Tuple<int, int, char>(tmp.Item1, tmp.Item2, data[3]);
        }

        //// parsed_to_model_coordinates takes a parsed coordinates string and returns
        //// as multiple values the row, col in terms of the model used by goboard.py.
        //// This assumes coords is "<letter><letter>" and valid indexes.
        ////
        //// SGF format is col,row (count from left edge, count from top).
        ////
        public static Tuple<int, int> ParsedToModelCoordinates (string coords) {
            if (coords == "")
                // Pass move
                return new Tuple<int, int>(GoBoardAux.NoIndex, GoBoardAux.NoIndex);
            else {
                coords = coords.ToLower();
                return new Tuple<int, int>(Array.IndexOf(GoBoardAux.letters, coords[1]),
                                           Array.IndexOf(GoBoardAux.letters, coords[0]));
            }
        }

        public static char ModelCoordinateToDisplayLetter (int i) {
            var chr_offset = (i < 9) ? i : (i + 1);
            var chr_txt = (char)(chr_offset + (int)('A') - 1);
            return chr_txt;
        }

        public static int DisplayLetterToModelCoordinate (char c) {
            c = char.ToLower(c);
            var i = c - 'a';
            if (i >= 9)
                return i;
            else
                return i + 1;
        }

    } // GoBoardAux class



    //// IMoveNext exists to abstract property access across Move and ParsedNodes for computing
    //// the layout of the visual tree of game moves.  We need this first class named type because
    //// UWP has made C# dynamic so slow it is useless (computing a tree layout went from milliseconds
    //// to taking 5s (debug on .net core) and 8s (release on .net native).
    ////
    //// These cannot be implicitly implemented and have distinct names because 1) collections cannot
    //// implicitly convert to collections of supertypes, 2) Next needs to return a Move when we need
    //// all the move functionality, and 3) Color doesn't exist on ParsedNode naturally.
    ////
    public interface IMoveNext {
        Color IMNColor { get; }
        IMoveNext IMNNext { get; }
        List<IMoveNext> IMNBranches { get; }
    }


    //// Move represents moves that have been rendered on the go board, or moves created for next moves
    //// of a current move, so that the current move can point to Moves.  These next moves are marked
    //// Rendered==false.
    ////
    public class Move : IMoveNext {
        public int Row { get; set; }  // These count from top to bottom, 1 based.
        public int Column { get; set; }  // These count from left to right, 1 based.
        public Color Color { get; set; }
        public Move Previous { get; set; }
        public Move Next { get; set; }
        public int Number { get; set; }
        public bool IsPass { get; set; }  // True when row, col are both GoBoardAux.NoIndex
        // DeadStones is always a list.
        public List<Move> DeadStones { get; set; }
        public List<Adornments> Adornments { get; set; }
        public string Comments { get; set; }
        // Branches is null when there is zero or one next move.
        public List<Move> Branches { get; set; }
        public ParsedNode ParsedNode { get; set; }
        public bool Rendered { get; set; }

        public Move(int row, int col, Color color) {
            this.Row = row;
            this.Column = col;
            this.IsPass = (row == GoBoardAux.NoIndex) && (col == GoBoardAux.NoIndex);
            this.Color = color;
            // The following need to be set if placing move into a game
            this.Next = null;
            this.Previous = null;
            this.Number = 0;
            this.DeadStones = new List<Move>();
            // Adornments is for editing markup.
            this.Adornments = new List<Adornments>();
            // Comments holds text describing the move or current game state.
            this.Comments = "";
            // Branches holds all the next moves, while next points to one of these.
            // This is None until there's more than one next move.
            this.Branches = null;
            // parsed_node is a node from a .sgf file.  This holds properties we may ignore, but
            // we include them when generating a file/persistence representation of moves, merging
            // with an state that's changed from user edits.
            this.ParsedNode = null;
            // rendered by default is true assuming moves are made and immediately
            // displayed, but parsed nodes can have unprocessed branches or
            // annotations.
            this.Rendered = true;
        }

        /// Thes IMN... member need to exist because C# dynamic is unusable in UWP (see IMoveNext type).
        /// 
        public IMoveNext IMNNext { get { return this.Next; } }
        public List<IMoveNext> IMNBranches { get { return (this.Branches == null) ? null : new List<IMoveNext>(this.Branches); } }
        public Color IMNColor { get { return this.Color; } }


        public Adornments AddAdornment (Adornments a) {
            this.Adornments.Add(a);
            return a;
        }

        public void RemoveAdornment(Adornments a) {
            this.Adornments.Remove(a);
        }

    } // Move class



    public enum AdornmentKind {
        Triangle, Square, Letter, CurrentMove
    }



    //// Adornments models adornments on stones or the board.  It contains a cookie to point
    //// to a UIElement to aid in removing or managing the UI manifestation of the adornment.
    ////
    public class Adornments {
        public AdornmentKind Kind { get; set; }
        public int Row { get; set; }
        public int Column { get; set; }
        public Move Move { get; set; }
        // Letter is string when kind is letter.
        public string Letter { get; set; }
        // Cookie is for the UI layer to hold onto objects per adornment.
        public UIElement Cookie { get; set; }

        public Adornments (AdornmentKind kind, int row, int col, Move move = null, string letter = null) {
            this.Kind = kind;
            this.Letter = letter;
            this.Row = row;
            this.Column = col;
            this.Move = move;
            this.Cookie = null;
        }

        //// There's only one current move adornment.
        ////
        private static Adornments _currentAdornment = new Adornments(AdornmentKind.CurrentMove, 1, 1);
        public static Adornments CurrentMoveAdornment {
            get { return _currentAdornment; }
        }

        //// GetCurrentMove reserves or locks the current move adornment.
        ////
        public static void GetCurrentMove (Move move, UIElement cookie) {
            // This debug.assert could arguably be a throw if we think of this function
            // as platform/library.
            MyDbg.Assert(Adornments.CurrentMoveAdornment.Move == null,
                         "Already have current move adornment at row, col: "); // +
                //Adornments.CurrentMoveAdornment.Move.Row.ToString() + ", " +
                //Adornments.CurrentMoveAdornment.Move.Column.ToString() + ".");
            Adornments.CurrentMoveAdornment.Move = move;
            Adornments.CurrentMoveAdornment.Cookie = cookie;
            move.AddAdornment(Adornments.CurrentMoveAdornment);
        }

        //// ReleaseCurrentMove releases the current move adornment so that it can be
        //// acquired to apply to another move.
        ////
        public static void ReleaseCurrentMove() {
            var move = Adornments.CurrentMoveAdornment.Move;
            // This debug.assert could arguably be a throw if we think of this function
            // as platform/library.
            MyDbg.Assert(move != null, "Do not have current move adornment.");
            move.Adornments.Remove(Adornments.CurrentMoveAdornment);
            Adornments.CurrentMoveAdornment.Move = null;
            Adornments.CurrentMoveAdornment.Cookie = null;
        }

    } // Adornments class

} // namespace
